execution_engine/src/runtime/externals.rs

```diff
@@ -1124,6 +1125,100 @@ where
 
                 Ok(Some(RuntimeValue::I32(api_error::i32_from(result))))
             }
+            FunctionIndex::ManageMessageTopic => {
+                // args(0) = pointer to the serialized topic name string in wasm memory
+                // args(1) = size of the serialized topic name string in wasm memory
+                // args(2) = pointer to the operation to be performed for the specified topic
+                // args(3) = size of the operation
+                let (topic_name_ptr, topic_name_size, operation_ptr, operation_size) =
+                    Args::parse(args)?;
+                self.charge_host_function_call(
+                    &host_function_costs.manage_message_topic,
+                    [
+                        topic_name_ptr,
+                        topic_name_size,
+                        operation_ptr,
+                        operation_size,
+                    ],
+                )?;
+
+                let limits = self.context.engine_config().wasm_config().messages_limits();
+
+                if topic_name_size > limits.max_topic_name_size() {
+                    return Ok(Some(RuntimeValue::I32(api_error::i32_from(Err(
+                        ApiError::MaxTopicNameSizeExceeded,
+                    )))));
+                }
+
+                let topic_name = self.string_from_mem(topic_name_ptr, topic_name_size)?;
```

Comment:
I know we're not consistent at all with regards to this but I have suggestion that we perhaps don't deserialize a string here, but instead, pass the string directly from Wasm. The host already has access to all the data, and you just need to pass pointer to the string i.e.

```rust
// wasm
fn foo(s: &str) {
  ext_ffi::foo(s.as_ptr(), s.len());
}
// host
self.checked_memory_slice(key_ptr as usize, key_size as usize, |data| {
  std::str::from_utf8(data).map(ToOwned::to_owned)
})?
```

This should make it cheaper for the users as additional serialization, deserialization, etc, adds to the gas costs.

Reply:
Removed serialization in https://github.com/casper-network/casper-node/pull/4370/commits/53483e16590b67de9be9be4d9ba4e6b64d1f22ee although didn't use `checked_memory_slice` as it's not available yet on the feat-2.0 branch. Will do a follow-up commit after.